src/main/java/undo/UndoController.java

```diff
@@ -0,0 +1,84 @@
+package undo;
+
+import backend.Logic;
+import backend.resource.TurboIssue;
+import javafx.application.Platform;
+import javafx.geometry.Insets;
+import javafx.scene.control.Label;
+import javafx.util.Pair;
+import ui.NotificationController;
+import ui.components.Notification;
+import undo.actions.Action;
+import util.DialogMessage;
+
+import java.util.Optional;
+
+public class UndoController {
+
+    @SuppressWarnings("SpellCheckingInspection")
+    private static final String OCTICON_INFO = "\uf059";
+
+    private final Logic logic;
+    private final NotificationController notificationController;
+    private Optional<Pair<TurboIssue, Action<TurboIssue>>> undoBuffer;
+
+    // Holds an undo buffer of size 1.
+    // Buffer stores the undo action.
+    public UndoController(Logic logic, NotificationController notificationController) {
+        this.logic = logic;
+        this.notificationController = notificationController;
+        undoBuffer = Optional.empty();
+    }
+
+    // Gets called when an action is undone
+    public void undoCallback() {
+        if (undoBuffer.isPresent()) {
+            determineLogicMethod(undoBuffer.get().getKey(), undoBuffer.get().getValue());
+            undoBuffer = Optional.empty();
+        }
+    }
+
+    // Adds an action to the undo buffer.
+    @SuppressWarnings("unchecked")
+    public void addAction(TurboIssue issue, Action action) {
+        determineLogicMethod(issue, action);
+        undoBuffer = Optional.of(new Pair<>((TurboIssue) action.act(issue), action.invert()));
+        Notification notification = new Notification(createInfoOcticon(),
+                "Undo " + action.getDescription() + " for #" + issue.getId() + ": " + issue.getTitle(),
+                "Undo", this::undoCallback);
+        notificationController.showNotification(notification);
+    }
+
+    // Check to determine which logic method to call
+    @SuppressWarnings("unchecked")
+    private void determineLogicMethod(TurboIssue issue, Action action) {
+        if (action.getType() == Action.ActionType.CHANGELABELS) {
+            logic.replaceIssueLabels(issue, action)
```

Comment:
Hmm, good point. It wouldn't make sense to do `action.act(a).undo(b)`.

I think it's fine to call the Logic method directly.

Depends if you need to wait on the result of the undo and do something on completion, etc. If you need to wait but don't need the result it can be `CompletableFuture<Void>`, otherwise `void` is okay. I don't think the two methods need to be symmetric because we don't have redo.

If we do this, we may not need `invert` either. Just put the action on the buffer and call `undo` there.